TeamCode/src/main/java/com/acmerobotics/relicrecovery/motion/MotionProfileGenerator.java

```diff
@@ -0,0 +1,202 @@
+                    // well if we are already slamming on the brakes then we might as well do infinite jerk
+                    //find out how much acceleration it will take
+                    double stoppingA = (Math.pow(goal.maxAbsV, 2) - Math.pow(profile.end().v, 2)) / (2 * (goal.pos - profile.end().x));
+                    double stoppingTime = (goal.maxAbsV - profile.end().v) / stoppingA;
+                    profile.appendInfiniteJerkControl(stoppingA, stoppingTime);
+                    return profile;
+                }
+                else if(constraints.endBehavior == MotionConstraints.END_BEHAVIOR.VIOLATE_MAX_ABS_V) {
+                    //we will just work with what we got - we will slow down as much as possible
+                    profile.appendControl(-constraints.maxJ, stoppingTimes[0]);
+                    profile.appendControl(0, stoppingTimes[1]);
+                    profile.appendControl(constraints.maxJ, stoppingTimes[2]);
+                    double timeAtGoal = profile.timeAtPos(goal.pos);
+                    return profile.trimAfter(timeAtGoal);
+                }
+            }
+        }
+
+        //ok cool we should be able to do it now
+
+        double dX = goal.pos - profile.end().x;
+        double j = constraints.maxJ;
+
+        //segments 1-3
+        double [] timesToAccel = getDeltaVTimes(constraints.maxV - profile.end().v, constraints);
+        //segments 5-7
+        double [] timesToDecel = getDeltaVTimes(goal.maxAbsV - constraints.maxV, constraints);
+
+        //figure out how much we are going to have to coast (segment 4)
+        MotionProfile accelProfile = new MotionProfile(new MotionState (0,0,0,0,0));
+        accelProfile.appendControl(j, timesToAccel[0]);
+        accelProfile.appendControl(0, timesToAccel[1]);
+        accelProfile.appendControl(-j, timesToAccel[2] + timesToDecel[0]);
+        accelProfile.appendControl(0, timesToDecel[1]);
+        accelProfile.appendControl(j, timesToDecel[2]);
+        double dxWithoutCoast = accelProfile.end().x;
+
+        //ok cool we can figure out how far to coast, but this only works if we have enough time to get up to max v
+        double dxCoast = dX - dxWithoutCoast;
+        double timeCoast = Math.max(0, dxCoast / constraints.maxV);
+
+        //now we need to solve the case where we can not get up to max v
+
+        //spend hours trying to figure out the correct way to find max reachable velocity
+        //sculpt velocity graphs out of mashed potatoes
+        //loose sleep over it
+        //give up and decide a binary search is not that bad
+
+        double maxV = constraints.maxV;
+        double epsilon = 1E-10;
+        double maxVmax = constraints.maxV;
+        double maxVmin = 0;
+        if (dxCoast < 0) {
+
+            //yeah its a hack but its so much easier than the alternative, and on the plus side it converges fairly fast and could definitely be optimized
+            //todo optimise this
```

Comment:
I am not sure how we could make better gueses for upper and lower bounds to lower the search space